### ISSUE TITLE
[Question] The number of users and items mismatch

### ISSUE THREAD

**Issue by trannhuthuat96** [RUCAIBox/RecBole#1079]
**The number of users and items does not match with reported numbers in paper**
1. When I run the code on datasets such as LastFM and Gowalla, I found that the number of ratings mismatch with the ones reported in paper https://arxiv.org/pdf/2011.01731.pdf. (The processed data is taken from shared Google Drive https://recbole.io/docs/user_guide/data/dataset_download.html)
E.g. In the log output, the number of ratings in Gowalla just ~4M, while the reported number is >6M

2. The number of users and items in the log output is +1 compared with ones reported in papers? Is this from padded item?

Could you help me to point out these?

**User and item IDs when splitting data**
When splitting data using Ratio Split or Leave-One-Out, some items will not appear in train data
3. Will these items added into item2id? For example, these items have embedding in item embedding matrix?  

4. When evaluating metrics such as Recall or Hit Ratio on validation or test set, will the results of these items be reported? Is there any APIs such that items that do not appear in training data will not be reported when evaluate on validation and test set?

Thanks,

**Comment by Sherry-XLL** [comment]
@trannhuthuat96 Hello, thanks for your attention to RecBole! 

- Firstly, the processed data set provided by RecBole retains the timestamp and ratings, and the k-core filtering steps of different papers and the original data set are not completely consistent. If you want to reproduce the conclusions in the paper, you'd better use the data set sources mentioned in the paper and process them into the atomic files required by RecBole.
- Secondly, we add `[PAD]` for all the token like fields. Thus after remapping ID, `0` will be reserved for `[PAD]`, which makes the result of `Dataset.item_num` more than the actual number. Note that for Knowledge-based models, we add one more relation called `U-I Relation`. It describes the history interactions which will be used in `recbole.data.dataset.kg_dataset.KnowledgeBasedDataset.ckg_graph()`. Thus the result of `KGDataset.relation_num` is two more than the actual number of relations.
- Thirdly, cold start items that do not appear in the training data cannot be recommended under normal conditions. Therefore, `item2id` contains all items that appear in the whole dataset.
- Last but not least, the results of these items will be reported, and there is no ready-made API to remove the impact of these items on the evaluation. Generally speaking, items have usually appeared in the training set, and you can implement a function to remove these items if you need one.

**Comment by Sherry-XLL** [comment]
@trannhuthuat96 For the first point, Gowalla is from a location-based social networking website where users share their locations by checking-in, and contains a total of 6,442,890 check-ins of these users over the period of Feb. 2009 - Oct. 2010. There are more than 6 million interactions in the raw dataset. After the duplicate interactions are deleted, there are more than four million ratings left. You can find the relevant raw data of Gowalla on the [official website](https://snap.stanford.edu/data/loc-gowalla.html).

For the third point, it is very unlikely that the items left one out will only appear in the validation and test set, so we did not consider this problem when developing. If you think this will affect the model results, you can delete the cold start item by setting `item_inter_num_interval (str)`, and items whose number of interactions is in the interval will be retained.

**Comment by trannhuthuat96** [comment]
@Sherry-XLL Thanks for your response.

For leave-one-out splitting, just setting item_inter_num_interval, let say [3, inf), can guarantee that at least one interaction appears in train/valid/test set. However, when considering ration-based split, it's hard to measure minimal interactions to filter out cold start problems.

Just want clarify your k-core processing mentioned in your CIKM-21 paper. Probably relates to `item_inter_num_interval (str)`. K-core filtering means recursively filter out item/user until their interactions in the processed data is K. Am I right? 
Or K-core filtering just remove items/users with less than K interactions in one pass then stop. This means in the processed data, some users/items may have less than K interactions. 

Lastly, just a suggestion. It would be better if data released in RecBole has their versions. Following up researches will refer to the version they used because the original dataset can be updated overtime.

Thanks for time and consideration!

**Comment by Sherry-XLL** [comment]
@trannhuthuat96 Hello, the cold start items you mentioned may appear when using ratio-based method, but with the appropriate configuration, we can hardly consider its impact, which is also a common practice in academia. If you think this is a problem that should be carefully considered, you can divide the train/eval/test data set first, and then load them into RecBole by `benchmark_filename`.

Your second understanding of k-core filtering is correct. We only limit the number of users and items to a given range.

Finally, the data sets used in RecBole are public data sets related to the recommendation system. You can find the relevant data set description and original links in our [RecSysDatasets repo](https://github.com/RUCAIBox/RecSysDatasets). The same data set will not change. If necessary, we will add widely used data sets on this basis. For the time being, we will not add a version number to the data set.

**Comment by trannhuthuat96** [comment]
@Sherry-XLL Thank you for your detailed answer. I'll try to explore `benchmark_filename`. The link to your datasets in RecBole is helpful.

`We only limit the number of users and items to a given range.` This sentence makes me a bit confused. 
You limit the number of users/items or the number of interactions per user/item?. 
Take an example, when k=5, we have 5-core filtering. This means that in the dataset after processing, each user interacts with at least 5 different items and each item is interacted by at least 5 different users. Is this correct?

**Comment by Sherry-XLL** [comment]
@trannhuthuat96 Well, your understanding of k-core filtering is correct. RecBole configuration has changed in different versions. In the current version, we use `user_inter_num_interval` and `item_inter_num_interval` to filter interactions. You can refer to our [documentation](https://recbole.io/docs/user_guide/config/data_settings.html#filter-by-number-of-interactions) for more details about the settings. 

**Filter by number of interactions**

- ``user_inter_num_interval (str)`` : Has the interval format, such as ``[A,B]`` / ``[A,B)`` / ``(A,B)`` / ``(A,B]``,  where ``A`` and ``B`` are the endpoints of the interval and ``A <= B``. Users whose number of interactions is in the interval will be retained. Defaults to ``[0,inf)``.
- ``item_inter_num_interval (str)`` : Has the interval format, such as ``[A,B]`` / ``[A,B)`` / ``(A,B)`` / ``(A,B]``,  where ``A`` and ``B`` are the endpoints of the interval and ``A <= B``. Items whose number of interactions is in the interval will be retained. Defaults to ``[0,inf)``.

As a result, if you want to apply 5-core filtering in a dataset, you can just set configuration as follows.

```python
user_inter_num_interval: "[5,inf)"
item_inter_num_interval: "[5,inf)"
```

**Comment by trannhuthuat96** [comment]
@Sherry-XLL Thanks for your suggestions!
